proxy/src/coco.rs

```diff
@@ -0,0 +1,467 @@
+//! Abstractions and utilities for git interactions through the API.
+
+use std::collections::hash_map::DefaultHasher;
+use std::hash::{Hash, Hasher};
+use std::str::FromStr;
+
+use librad::git;
+use librad::keys;
+use librad::meta::{self, common::Url};
+use librad::paths::Paths;
+use librad::peer;
+use librad::project;
+use radicle_surf as surf;
+use radicle_surf::git::git2;
+
+use crate::error;
+
+/// Branch name representation.
+#[derive(Debug, Eq, Ord, PartialEq, PartialOrd)]
+pub struct Branch(pub String);
```

Comment:
Curious why we need `pub` visibility here, since this would allow branches to be constructed incorrectly, eg. with illegal characters from anywhere in the code. Same for `Tag`.

Reply:
Addressed by making it private and implementing `std::fmt::Display` for them.